pkgs/top-level/aliases.nix

```diff
@@ -209,15 +207,62 @@ let
   # Make sure that we are not shadowing something from all-packages.nix.
   checkInPkgs =
     n: alias:
-    if builtins.hasAttr n super then throw "Alias ${n} is still in all-packages.nix" else alias;
+    if builtins.hasAttr n _super then throw "Alias ${n} is still in all-packages.nix" else alias;
+
+  # `config.warnAliases` machinery
+  # These are no-ops when warnAliases=false
+
+  selfWithAliasWarningNames =
+    if _super.config.warnAliases then
+      lib.mapAttrs (
+        name: package:
+        if lib.isDerivation package then
+          # smuggle the attribute name of each package
+          package // { __aliasWarningAttrName = name; }
+        else
+          package
+      ) self
+    else
+      self;
+
+  addAliasWarnOnInstantiate =
+    if _super.config.warnAliases then
+      name: alias:
+      if !alias ? __aliasWarningAttrName then
+        alias
+      else
+        # our lib.warnOnInstantiate override also removes __aliasWarningAttrName
+        lib.warnOnInstantiate "'${name}' has been renamed to/replaced by '${alias.__aliasWarningAttrName}'"
+          alias
+    else
+      name: alias: alias;
+
+  lib = lib'.extend (
```

Comment:
use of extend was discouraged. cc @roberth. 
possibly extend lib without calling `extend`, 
I am a bit worried about this because after this PR:
`pkgs.lib` != `lib`

Maybe bind `__aliasWarningAttrName = { warnAliases = true; }` ? That would also allow perPackage decisions which by default inherit true from the config of pkgs ?
(Just playing with ideas here)

Reply:
How about renaming `lib` to `_lib` in the file with a comment about only using the "whitelisted" functions in a `inherit (_lib) ...;` in the let binding is allowed? The we're free to make custom versions of some of the lib functions.

> Maybe bind `__aliasWarningAttrName = { warnAliases = true; }` ? That would also allow perPackage decisions which by default inherit true from the config of pkgs ?
> (Just playing with ideas here)

I don't quite understand what you mean by this nor what it achieves.